In [1]:
!ls

sample_data/


<p style="align: center;"><img src="https://static.tildacdn.com/tild6636-3531-4239-b465-376364646465/Deep_Learning_School.png", width=600, height=300></p>

<h3 style="text-align: center;"><b>Физтех-Школа Прикладной математики и информатики (ФПМИ) МФТИ</b></h3>

<h2 style="text-align: center;"><b>Библиотека <a href="http://pandas.pydata.org/">pandas</a></b></h2>

Библиотека `pandas` активно используется в современном data science для работы с данными, которые могут быть представлены в виде таблиц (а это очень, очень большая часть данных)

`pandas` есть в пакете Anaconda, но если вдруг у Вас её по каким-то причинам нет, то можно установить, раскомментировав следующую команду:

In [ ]:
#!pip install pandas

In [2]:
import pandas as pd

Нужны данные. Про AppStore слышали? :)

### Скачать данные: https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps

Нас интересует файл `AppStore.csv`.

Кстати, `.csv` (Comma Separated Values) - наверное, самый частый формат данных в современном data science. По сути, это то же самое, что и `.xls` (и `.xlsx`), то есть таблица. 

Итак, *чтение файла с данными*:

In [3]:
data = pd.read_csv('AppleStore.csv')

Посмотрим, что же такое переменная `data`:

In [4]:
data

Unnamed: 0          id  ... lang.num  vpp_lic
0              1   281656475  ...       10        1
1              2   281796108  ...       23        1
2              3   281940292  ...        3        1
3              4   282614216  ...        9        1
4              5   282935706  ...       45        1
...          ...         ...  ...      ...      ...
7192       11081  1187617475  ...        1        1
7193       11082  1187682390  ...        1        1
7194       11087  1187779532  ...        1        1
7195       11089  1187838770  ...        2        1
7196       11097  1188375727  ...        2        1

[7197 rows x 17 columns]

In [5]:
type(data)

pandas.core.frame.DataFrame

Тип == `pandas.core.frame.DataFrame`, обычно говорят просто "датафрейм", то есть "кусок данных".

Можно вывести информацию по датафрейму:

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7197 entries, 0 to 7196
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        7197 non-null   int64  
 1   id                7197 non-null   int64  
 2   track_name        7197 non-null   object 
 3   size_bytes        7197 non-null   int64  
 4   currency          7197 non-null   object 
 5   price             7197 non-null   float64
 6   rating_count_tot  7197 non-null   int64  
 7   rating_count_ver  7197 non-null   int64  
 8   user_rating       7197 non-null   float64
 9   user_rating_ver   7197 non-null   float64
 10  ver               7197 non-null   object 
 11  cont_rating       7197 non-null   object 
 12  prime_genre       7197 non-null   object 
 13  sup_devices.num   7197 non-null   int64  
 14  ipadSc_urls.num   7197 non-null   int64  
 15  lang.num          7197 non-null   int64  
 16  vpp_lic           7197 non-null   int64  


И более статистически значимую информацию:

In [7]:
data.describe()

Unnamed: 0            id  ...     lang.num      vpp_lic
count   7197.000000  7.197000e+03  ...  7197.000000  7197.000000
mean    4759.069612  8.631310e+08  ...     5.434903     0.993053
std     3093.625213  2.712368e+08  ...     7.919593     0.083066
min        1.000000  2.816565e+08  ...     0.000000     0.000000
25%     2090.000000  6.000937e+08  ...     1.000000     1.000000
50%     4380.000000  9.781482e+08  ...     1.000000     1.000000
75%     7223.000000  1.082310e+09  ...     8.000000     1.000000
max    11097.000000  1.188376e+09  ...    75.000000     1.000000

[8 rows x 12 columns]

Итак, наши данные -- это какая-то **информация про приложения в AppStore**. Давайте посмотрим, что тут есть:

In [8]:
# все столбцы датафрейма (названия)
data.columns

Index(['Unnamed: 0', 'id', 'track_name', 'size_bytes', 'currency', 'price',
       'rating_count_tot', 'rating_count_ver', 'user_rating',
       'user_rating_ver', 'ver', 'cont_rating', 'prime_genre',
       'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'],
      dtype='object')

---

#### Задание 0

О чём данные? (судя по столбцам)

---

Все значения датафрейма:

In [ ]:
# одна строка == описание одного объекта, в данном случае -- приложения из AppStore
data.values

array([[1, 281656475, 'PAC-MAN Premium', ..., 5, 10, 1],
       [2, 281796108, 'Evernote - stay organized', ..., 5, 23, 1],
       [3, 281940292, 'WeatherBug - Local Weather, Radar, Maps, Alerts',
        ..., 5, 3, 1],
       ...,
       [11087, 1187779532, 'Bret Michaels Emojis + Lyric Keyboard', ...,
        1, 1, 1],
       [11089, 1187838770, 'VR Roller Coaster World - Virtual Reality',
        ..., 0, 2, 1],
       [11097, 1188375727, 'Escape the Sweet Shop Series', ..., 0, 2, 1]],
      dtype=object)

In [ ]:
data.values[0]

array([1, 281656475, 'PAC-MAN Premium', 100788224, 'USD', 3.99, 21292, 26,
       4.0, 4.5, '6.3.5', '4+', 'Games', 38, 5, 10, 1], dtype=object)

In [ ]:
data.shape

(7197, 17)

-- то есть 7197 строк и 17 столбцов.

Важно уметь обращаться к определённой строке или столбцу dataframe'а. Индексирование очень похоже на numpy, однако есть свои тонкости:

* Получить весь столбец (в даном случае столбец 'track_name'):

In [ ]:
data['track_name']

0                                         PAC-MAN Premium
1                               Evernote - stay organized
2         WeatherBug - Local Weather, Radar, Maps, Alerts
3       eBay: Best App to Buy, Sell, Save! Online Shop...
4                                                   Bible
5                                        Shanghai Mahjong
6                  PayPal - Send and request money safely
7                                 Pandora - Music & Radio
8                             PCalc - The Best Calculator
9                                             Ms. PAC-MAN
10                              Solitaire by MobilityWare
11                                       SCRABBLE Premium
12                   Google – Search made just for mobile
13                       Bank of America - Mobile Banking
14                                               FreeCell
15                 TripAdvisor Hotels Flights Restaurants
16                                               Facebook
17         Yel

In [ ]:
type(data['track_name'])

pandas.core.series.Series

In [ ]:
data['track_name'][0]

'PAC-MAN Premium'

`pandas.core.series.Series` -- это тип подвыборки датафрейма. С этим типом работают реже, но стоит знать, что он есть.

* Получить определённую строку:

In [ ]:
data.iloc[10]

Unnamed: 0                                 11
id                                  284791396
track_name          Solitaire by MobilityWare
size_bytes                           49618944
currency                                  USD
price                                    4.99
rating_count_tot                        76720
rating_count_ver                         4017
user_rating                               4.5
user_rating_ver                           4.5
ver                                    4.10.1
cont_rating                                4+
prime_genre                             Games
sup_devices.num                            38
ipadSc_urls.num                             4
lang.num                                   11
vpp_lic                                     1
Name: 10, dtype: object

Более продвинутое и полезное индексирование:

In [ ]:
data.iloc[5, 10]

'1.8'

-- `data.iloc[i, j]`, где `i` -- номер строки, `j` -- номер столбца

Почти все типы индексирования:

* к строкам по числовому индексу, к столбцам -- по имени

In [ ]:
data.loc[[0, 1, 2], ['track_name', 'id']]

track_name         id
0                                  PAC-MAN Premium  281656475
1                        Evernote - stay organized  281796108
2  WeatherBug - Local Weather, Radar, Maps, Alerts  281940292

* к строкам и к столбцам по числовому индексу

In [ ]:
data.iloc[[1, 3, 5], [2, 1]]

track_name         id
1                          Evernote - stay organized  281796108
3  eBay: Best App to Buy, Sell, Save! Online Shop...  282614216
5                                   Shanghai Mahjong  283619399

* поддерживает и то, и то, но **запрещён**:

In [ ]:
data.ix[[2,1,0], ['track_name', 'id']]

c:\users\izakharkin\anaconda3\envs\vision\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


track_name         id
2  WeatherBug - Local Weather, Radar, Maps, Alerts  281940292
1                        Evernote - stay organized  281796108
0                                  PAC-MAN Premium  281656475

In [ ]:
data.ix[[0,1,2], [2, 1]]

c:\users\izakharkin\anaconda3\envs\vision\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


track_name         id
0                                  PAC-MAN Premium  281656475
1                        Evernote - stay organized  281796108
2  WeatherBug - Local Weather, Radar, Maps, Alerts  281940292

Точно также дело обстоит со **срезами**:

In [ ]:
data.iloc[5:100, 0:5]

Unnamed: 0         id                                         track_name  \
5            6  283619399                                   Shanghai Mahjong   
6            7  283646709             PayPal - Send and request money safely   
7            8  284035177                            Pandora - Music & Radio   
8            9  284666222                        PCalc - The Best Calculator   
9           10  284736660                                        Ms. PAC-MAN   
10          11  284791396                          Solitaire by MobilityWare   
11          12  284815117                                   SCRABBLE Premium   
12          13  284815942               Google – Search made just for mobile   
13          14  284847138                   Bank of America - Mobile Banking   
14          15  284862767                                           FreeCell   
15          16  284876795             TripAdvisor Hotels Flights Restaurants   
16          17  284882215                                           Facebook   
17          18  284910350     Yelp - Nearby Restaurants, Shopping & Services   
18          20  284993459  Shazam - Discover music, artists, videos & lyrics   
19          21  285005463                      Crash Bandicoot Nitro Kart 3D   
20          22  285946052                                             iQuran   
21          23  285994151                                        :) Sudoku +   
22          24  286058814    Yahoo Sports - Teams, Scores, News & Highlights   
23          25  286070473                          Mileage Log | Fahrtenbuch   
24          27  286799607                        Cleartune - Chromatic Tuner   
25          28  286906691          Lifesum – Inspiring healthy lifestyle app   
26          29  286911400                                           Hangman.   
27          31  288113403      iTranslate - Language Translator & Dictionary   
28          32  288120394                                           TouchOSC   
29          33  288419283                                         RadarScope   
30          34  288429040                                           LinkedIn   
31          35  289084315                              Period Tracker Deluxe   
32          36  289446241                                  Election 2016 Map   
33          37  289523017                          Blackjack by MobilityWare   
34          39  289894882                                        White Noise   
..         ...        ...                                                ...   
70          80  299853944                                新浪新闻-阅读最新时事热门头条资讯视频   
71          81  299949744                                        MotionX GPS   
72          82  300048137                     AccuWeather - Weather for Life   
73          83  300238550      Mint: Personal Finance, Budget, Bills & Money   
74          84  300255638            ABC News - US & World News + Live Video   
75          85  300265786                    Rowmote: Remote Control for Mac   
76          86  300590611                                      Peekaboo Barn   
77          87  300704847                                 Speedtest by Ookla   
78          89  301387274                                         Pocket God   
79          90  301521403     Nike+ Training Club - Workouts & Fitness Plans   
80          91  301724680                                       Citi Mobile®   
81          92  302049354                  Phase 10 Pro - Play Your Friends!   
82          93  302584613        Kindle – Read eBooks, Magazines & Textbooks   
83          95  303058767                                               MASH   
84          96  303191318                                          同花顺-炒股、股票   
85          97  303337299                                Wedding Dash Deluxe   
86          99  303849934                                     Beer Pong Game   
87         100  304731501                    Tempo - Metronome with Setlists   
88  

Но обычно нужно уметь отвечать на более осмысленные вопросы, например:

* сколько приложений имеют рейтинг юзеров >= 4?

In [ ]:
len(data[data['user_rating'] >= 4])

4781

* какие есть валюты в столбце currency?

In [ ]:
np.unique(data['currency'])

array(['USD'], dtype=object)

* сколько приложений жанра Games? А тех, которые Games и рейтинг >= 4?

In [ ]:
# 1
len(data[data['prime_genre'] == 'Games'])

3862

In [ ]:
# 2
len(data[(data['user_rating'] >= 4) & (data['prime_genre'] == 'Games')])

2868

Оператор "ИЛИ" аналогично &, только он через |

In [ ]:
len(data[(data['user_rating'] >= 4) | (data['prime_genre'] == 'Games')])

5775

---

#### Задание 1

Получите ответ на следующие вопросы (с помощью pandas):

1). Сколько всего жанров и какие они?  
2). Сколько суммарно байт весят все приложения жанра Finance?  
3). Сколько всего бесплатных приложений, у которых рейтинг юзеров больше трёх?  
4). Есть ли приложения, которые стоят больше 100 долларов?

In [ ]:
# Ваш код здесь

---

Часто надо сделать что-то такое:

In [ ]:
import numpy as np

In [ ]:
np.array(data['size_bytes'])

array([100788224, 158578688, 100524032, ..., 111322112,  97235968,
        90898432], dtype=int64)

То есть взять столбец из таблицы и засунуть его в np.array(). Это можно сделать и просто обратившись к .values, ведь его тип и есть np.array:

In [ ]:
type(data['size_bytes'].values)

numpy.ndarray

Больше про pandas можно найти по этом полезным ссылкам:

* Официальные туториалы: http://pandas.pydata.org/pandas-docs/stable/tutorials.html

* Статья на Хабре от OpenDataScience сообщества: https://habr.com/company/ods/blog/322626/

* Подробный гайд: https://media.readthedocs.org/pdf/pandasguide/latest/pandasguide.pdf

Главное в работе с новыми библиотеками -- не бояться тыкать в разные функции, смотреть типы возвращаемых объектов и активно пользоваться Яндексом, а ещё лучше понимать всё из docstring'а (`Shift+Tab` при нахождении курсора внутри скобок функции).